api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto

```diff
@@ -895,6 +895,14 @@ message HttpConnectionManager {
   // This should be set to ``false`` in cases where Envoy's view of the downstream address may not correspond to the
   // actual client address, for example, if there's another proxy in front of the Envoy.
   google.protobuf.BoolValue add_proxy_protocol_connection_state = 53;
+
+  // The timeout seconds configured here will be set in the "Keep-Alive" response header.
+  // For example, configuring 10s will return the response header "Connection: keep-alive" and "Keep-Alive: timeout=10".
+  // If not specified, the default is 0, which means this behavior is disabled.
+  // The "Keep-Alive" header field is recognized by Mozilla and Apache HTTPClient.
+  // Note that the "Connection" and "Keep-Alive" response headers will only be added when the downstream protocol is HTTP1.0 or HTTP1.1
+  // and it is not an upgrade connection scenario.
+  google.protobuf.Duration keepalive_header_timeout = 58;
```

Comment:
instead of adding a new field, WDYT of using  the already existing 
https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/protocol.proto#envoy-v3-api-field-config-core-v3-httpprotocoloptions-max-connection-duration

Reply:
hah, good point

Reply:
I do not prefer to use this configuration for two reasons: 
1. This configuration addresses the need for overall connection keep-alive time, but currently there is no such requirement for this issue (and enabling this configuration would result in increased new connections, which we do not want to have as a side effect of resolving this issue). Therefore, if we really need to reuse the configuration, it may be more appropriate to use idle_timeout instead; 
2. We should not consider reusing all existing server-side timeout configurations, whether it is idle_timeout or max_connection_duration, because if this timeout time is communicated to the client through the keep-alive header, the client's perception of when the connection will be closed may not truly synchronize with that of the server and there may be a race condition. For example, if the server reaches its timeout time and closes the connection by sending a FIN packet to the client, but at that moment the client has not yet reached its timeout time and has also not received the FIN packet yet, then if it continues sending requests on that connection, those requests will fail.

So I think the design of this `keepalive_timeout` [directive](https://nginx.org/en/docs/http/ngx_http_core_module.html#keepalive_timeout) in Nginx is reasonable, separating the configuration into two timeout values. One is used to control the idle timeout for connections with Downstream, and the other is used to set the keep-alive response header. This allows setting a relatively small value to ensure that clients close the connection before the server does.

Reply:
Any server/scenario is likely to have a race condition as you describe.
If all you want is a forced arbitrary keep alive header which has no relation to separately configured idle timeouts (a configuration option which is most likely going to result in misconfiguration and sub-optimal perfornace) I suggest you use the already existing response_headers_to_add rather than adding a new field.
/wait-any

Reply:
Connection and Keep-Alive are hop-by-hop headers. Envoy will clean them up in the EncodeHeader phase of HCM, so it is ineffective to add them through response_headers_to_add, etc.

Reply:
We should allow users to configure a value smaller than the idle timeout, in order to ensure that the Downstream closes the connection before Envoy does. As for the issue of misconfiguration, I think we can perform some validation during the configuration phase to restrict setting values larger than the idle timeout. What do you think?

Reply:
I'd strongly prefer we find a way to reuse existing fields rather than add more config here, especially as I think we're going to move hop by hop sanitization closer to the codecs as part of UHV.

We could for example special-case keep-alive in the router code, so if we see it we add an x-envoy-keep-alive and mutate that back to connection keep alive as part of mutate request headers.

cc @yanavlasov for other ideas
/wait-any